config dump: type map[string]interface {} is not assignable to type platforms.Matcher

### Description

Go runtime panics when I try to call dump a configuration by `contianerd config dump`.  Version 1.7.0-rc0.


### Steps to reproduce the issue

1. `containerd config default > test.config`
2. `containerd --config test.config config dump`
3. `containerd --config test.config` (this would also produce a similar error)


### Describe the results you received and expected

```
sudo ./containerd --config ../default.config config dump
panic: reflect.Set: value of type map[string]interface {} is not assignable to type platforms.Matcher

goroutine 1 [running]:
reflect.Value.assignTo({0x560d6092d480?, 0xc0003da990?, 0x7ff7bc7211c8?}, {0x560d604e8c2c, 0xb}, 0x560d60969120, 0xc0005522a0)
        /usr/local/go/src/reflect/value.go:3245 +0x2a5
reflect.Value.Set({0x560d60969120?, 0xc0005522a0?, 0x560d60969120?}, {0x560d6092d480?, 0xc0003da990?, 0xc0005205d0?})
        /usr/local/go/src/reflect/value.go:2236 +0xeb
github.com/pelletier/go-toml.(*Decoder).valueFromTree(0xc0003f4490, {0x560d60bda000, 0x560d60ac7140}, 0xc0003da600, 0x0)
        /go/src/github.com/containerd/containerd/vendor/github.com/pelletier/go-toml/marshal.go:811 +0x1a88
github.com/pelletier/go-toml.(*Decoder).valueFromTreeSlice(0xc0003f4490, {0x560d60bda000, 0x560d608495e0}, {0xc0005221f8, 0x1, 0xc0003f3aa8?})
        /go/src/github.com/containerd/containerd/vendor/github.com/pelletier/go-toml/marshal.go:916 +0x1bf
github.com/pelletier/go-toml.(*Decoder).valueFromToml(0xc0003f4490, {0x560d60bda000, 0x560d608495e0}, {0x560d60842960?, 0xc0005208b8?}, 0xc0005205a0)
        /go/src/github.com/containerd/containerd/vendor/github.com/pelletier/go-toml/marshal.go:1007 +0x736
github.com/pelletier/go-toml.(*Decoder).valueFromTree(0xc0003f4490, {0x560d60bda000, 0x560d60ac6f40}, 0xc0003da930, 0xc0003f4370)
        /go/src/github.com/containerd/containerd/vendor/github.com/pelletier/go-toml/marshal.go:807 +0x1a25
github.com/pelletier/go-toml.(*Decoder).unmarshal(0xc0003f4490, {0x560d60820ac0?, 0xc0003006c0})
        /go/src/github.com/containerd/containerd/vendor/github.com/pelletier/go-toml/marshal.go:736 +0x39f
github.com/pelletier/go-toml.(*Tree).Unmarshal(...)
        /go/src/github.com/containerd/containerd/vendor/github.com/pelletier/go-toml/marshal.go:621
github.com/containerd/containerd/services/server/config.(*Config).Decode(0xc0003ab340, 0xc000300720)
        /go/src/github.com/containerd/containerd/services/server/config/config.go:181 +0x245
github.com/containerd/containerd/cmd/containerd/command.outputConfig(0xc0003ab340)
        /go/src/github.com/containerd/containerd/cmd/containerd/command/config.go:63 +0x116
github.com/containerd/containerd/cmd/containerd/command.glob..func2(0xc000367b80?)
        /go/src/github.com/containerd/containerd/cmd/containerd/command/config.go:119 +0x7f
github.com/urfave/cli.HandleAction({0x560d608a2320?, 0x560d60ba2238?}, 0x4?)
        /go/src/github.com/containerd/containerd/vendor/github.com/urfave/cli/app.go:524 +0x50
github.com/urfave/cli.Command.Run({{0x560d604dd1cc, 0x4}, {0x0, 0x0}, {0x0, 0x0, 0x0}, {0x560d605610de, 0x48}, {0x0, ...}, ...}, ...)
        /go/src/github.com/containerd/containerd/vendor/github.com/urfave/cli/command.go:175 +0x67b
github.com/urfave/cli.(*App).RunAsSubcommand(0xc0003ab180, 0xc0003678c0)
        /go/src/github.com/containerd/containerd/vendor/github.com/urfave/cli/app.go:405 +0xe87
github.com/urfave/cli.Command.startApp({{0x560d604df709, 0x6}, {0x0, 0x0}, {0x0, 0x0, 0x0}, {0x560d6051ef19, 0x24}, {0x0, ...}, ...}, ...)
        /go/src/github.com/containerd/containerd/vendor/github.com/urfave/cli/command.go:380 +0xb7f
github.com/urfave/cli.Command.Run({{0x560d604df709, 0x6}, {0x0, 0x0}, {0x0, 0x0, 0x0}, {0x560d6051ef19, 0x24}, {0x0, ...}, ...}, ...)
        /go/src/github.com/containerd/containerd/vendor/github.com/urfave/cli/command.go:103 +0x845
github.com/urfave/cli.(*App).Run(0xc0003aafc0, {0xc0000520a0, 0x5, 0x5})
        /go/src/github.com/containerd/containerd/vendor/github.com/urfave/cli/app.go:277 +0xb87
main.main()
        github.com/containerd/containerd/cmd/containerd/main.go:39 +0x37
        ```

### What version of containerd are you using?

 containerd github.com/containerd/containerd v1.7.0-rc.0 2be87c1a75505ac066bec4a99e6c72aafd41d883

### Any other relevant information

```
> runc --version
runc version 1.1.4
commit: v1.1.4-0-g5fd4c4d
spec: 1.0.2-dev
go: go1.19.6
libseccomp: 2.5.1
```
```
> uname -a
Linux devbox 5.15.0-1030-aws #34~20.04.1-Ubuntu SMP Tue Jan 24 15:16:46 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
```

### Show configuration if it is related to CRI plugin.

_No response_

This issue is also causing containerd to crash on startup when run as a service under Windows. Here's the stacktrace reported in `C:\ProgramData\containerd\root\panic.log`:

```
panic: reflect.Set: value of type map[string]interface {} is not assignable to type platforms.Matcher

goroutine 42 [running]:
reflect.Value.assignTo({0x133a120?, 0xc0005538c0?, 0x274fd512628?}, {0x1570a98, 0xb}, 0x1369900, 0xc000551d50)
	/usr/local/go/src/reflect/value.go:3245 +0x2a5
reflect.Value.Set({0x1369900?, 0xc000551d50?, 0x1369900?}, {0x133a120?, 0xc0005538c0?, 0xc000540a98?})
	/usr/local/go/src/reflect/value.go:2236 +0xeb
github.com/pelletier/go-toml.(*Decoder).valueFromTree(0xc00044f7b8, {0x17d3350, 0x14a0840}, 0xc0002f7e60, 0x0)
	/go/src/github.com/containerd/containerd/vendor/github.com/pelletier/go-toml/marshal.go:811 +0x1a88
github.com/pelletier/go-toml.(*Decoder).valueFromTreeSlice(0xc00044f7b8, {0x17d3350, 0x12960e0}, {0xc00000a050, 0x1, 0x33?})
	/go/src/github.com/containerd/containerd/vendor/github.com/pelletier/go-toml/marshal.go:916 +0x1bf
github.com/pelletier/go-toml.(*Decoder).valueFromToml(0xc00044f7b8, {0x17d3350, 0x12960e0}, {0x12901e0?, 0xc0005403a8?}, 0xc000540a68)
	/go/src/github.com/containerd/containerd/vendor/github.com/pelletier/go-toml/marshal.go:1007 +0x736
github.com/pelletier/go-toml.(*Decoder).valueFromTree(0xc00044f7b8, {0x17d3350, 0x14a0640}, 0xc000553860, 0xc00044f698)
	/go/src/github.com/containerd/containerd/vendor/github.com/pelletier/go-toml/marshal.go:807 +0x1a25
github.com/pelletier/go-toml.(*Decoder).unmarshal(0xc00044f7b8, {0x1272b60?, 0xc000076780})
	/go/src/github.com/containerd/containerd/vendor/github.com/pelletier/go-toml/marshal.go:736 +0x39f
github.com/pelletier/go-toml.(*Tree).Unmarshal(...)
	/go/src/github.com/containerd/containerd/vendor/github.com/pelletier/go-toml/marshal.go:621
github.com/containerd/containerd/services/server/config.(*Config).Decode(0xc000409a40, 0xc0000767e0)
	/go/src/github.com/containerd/containerd/services/server/config/config.go:181 +0x245
github.com/containerd/containerd/services/server.New({0x17c0328, 0xc0004043c0}, 0xc000409a40)
	/go/src/github.com/containerd/containerd/services/server/server.go:224 +0x12ea
github.com/containerd/containerd/cmd/containerd/command.App.func1.1()
	/go/src/github.com/containerd/containerd/cmd/containerd/command/main.go:195 +0x9c
created by github.com/containerd/containerd/cmd/containerd/command.App.func1
	/go/src/github.com/containerd/containerd/cmd/containerd/command/main.go:192 +0x93b
```

I've tested with both 1.7.0-rc.0 and 1.7.0-beta.4, and can confirm that this does not reproduce with 1.7.0-beta.4.

--- a/pkg/transfer/local/transfer.go
+++ b/pkg/transfer/local/transfer.go
@@ -22,7 +22,6 @@ import (
 	"io"
 	"time"
 
-	"github.com/containerd/containerd"
 	"github.com/containerd/containerd/content"
 	"github.com/containerd/containerd/errdefs"
 	"github.com/containerd/containerd/images"
@@ -30,7 +29,6 @@ import (
 	"github.com/containerd/containerd/pkg/kmutex"
 	"github.com/containerd/containerd/pkg/transfer"
 	"github.com/containerd/containerd/pkg/unpack"
-	"github.com/containerd/containerd/platforms"
 	"github.com/containerd/typeurl/v2"
 	"golang.org/x/sync/semaphore"
 )
@@ -153,10 +151,9 @@ func (ts *localTransferService) withLease(ctx context.Context, opts ...leases.Op
 
 type TransferConfig struct {
 	// MaxConcurrentDownloads is the max concurrent content downloads for pull.
-	MaxConcurrentDownloads int `toml:"max_concurrent_downloads"`
-
+	MaxConcurrentDownloads int
 	// MaxConcurrentUploadedLayers is the max concurrent uploads for push
-	MaxConcurrentUploadedLayers int `toml:"max_concurrent_uploaded_layers"`
+	MaxConcurrentUploadedLayers int
 
 	// DuplicationSuppressor is used to make sure that there is only one
 	// in-flight fetch request or unpack handler for a given descriptor's
@@ -169,21 +166,8 @@ type TransferConfig struct {
 	BaseHandlers []images.Handler
 
 	// UnpackPlatforms are used to specify supported combination of platforms and snapshotters
-	UnpackPlatforms []unpack.Platform `toml:"unpack_platforms"`
+	UnpackPlatforms []unpack.Platform
 
 	// RegistryConfigPath is a path to the root directory containing registry-specific configurations
-	RegistryConfigPath string `toml:"config_path"`
-}
-
-func DefaultConfig() *TransferConfig {
-	return &TransferConfig{
-		MaxConcurrentDownloads:      3,
-		MaxConcurrentUploadedLayers: 3,
-		UnpackPlatforms: []unpack.Platform{
-			{
-				Platform:       platforms.Only(platforms.DefaultSpec()),
-				SnapshotterKey: containerd.DefaultSnapshotter,
-			},
-		},
-	}
+	RegistryConfigPath string
 }

--- a/plugins/transfer/plugin.go
+++ b/plugins/transfer/plugin.go
@@ -17,9 +17,14 @@
 package transfer
 
 import (
+	"fmt"
+
+	"github.com/containerd/containerd"
 	"github.com/containerd/containerd/leases"
 	"github.com/containerd/containerd/metadata"
 	"github.com/containerd/containerd/pkg/transfer/local"
+	"github.com/containerd/containerd/pkg/unpack"
+	"github.com/containerd/containerd/platforms"
 	"github.com/containerd/containerd/plugin"
 
 	// Load packages with type registrations
@@ -36,9 +41,9 @@ func init() {
 			plugin.LeasePlugin,
 			plugin.MetadataPlugin,
 		},
-		Config: local.DefaultConfig(),
+		Config: defaultConfig(),
 		InitFn: func(ic *plugin.InitContext) (interface{}, error) {
-			config := ic.Config.(*local.TransferConfig)
+			config := ic.Config.(*transferConfig)
 			m, err := ic.Get(plugin.MetadataPlugin)
 			if err != nil {
 				return nil, err
@@ -49,7 +54,57 @@ func init() {
 				return nil, err
 			}
 
-			return local.NewTransferService(l.(leases.Manager), ms.ContentStore(), metadata.NewImageStore(ms), config), nil
+			// Set configuration based on default or user input
+			var lc local.TransferConfig
+			lc.MaxConcurrentDownloads = config.maxConcurrentDownloads
+			lc.MaxConcurrentUploadedLayers = config.maxConcurrentUploadedLayers
+			for _, uc := range config.unpackConfiguration {
+				p, err := platforms.Parse(uc.platform)
+				if err != nil {
+					return nil, fmt.Errorf("%s: platform configuration %v invalid", plugin.TransferPlugin, uc.platform)
+				}
+
+				up := unpack.Platform{
+					Platform:       platforms.OnlyStrict(p),
+					SnapshotterKey: uc.snapshotter,
+				}
+				lc.UnpackPlatforms = append(lc.UnpackPlatforms, up)
+			}
+			lc.RegistryConfigPath = config.registryConfigPath
+
+			return local.NewTransferService(l.(leases.Manager), ms.ContentStore(), metadata.NewImageStore(ms), &lc), nil
 		},
 	})
+}
+
+type transferConfig struct {
+	// maxConcurrentDownloads is the max concurrent content downloads for pull.
+	maxConcurrentDownloads int `toml:"max_concurrent_downloads"`
+
+	// maxConcurrentUploadedLayers is the max concurrent uploads for push
+	maxConcurrentUploadedLayers int `toml:"max_concurrent_uploaded_layers"`
+
+	// unpackConfiguration is used to read config from toml
+	unpackConfiguration []unpackConfiguration `toml:"unpack_config"`
+
+	// registryConfigPath is a path to the root directory containing registry-specific configurations
+	registryConfigPath string `toml:"config_path"`
+}
+
+type unpackConfiguration struct {
+	platform    string
+	snapshotter string
+}
+
+func defaultConfig() *transferConfig {
+	return &transferConfig{
+		maxConcurrentDownloads:      3,
+		maxConcurrentUploadedLayers: 3,
+		unpackConfiguration: []unpackConfiguration{
+			{
+				platform:    platforms.Format(platforms.DefaultSpec()),
+				snapshotter: containerd.DefaultSnapshotter,
+			},
+		},
+	}
 }

